libutils/file_lib.c

```diff
@@ -537,19 +563,33 @@ int safe_open(const char *pathname, int flags, ...)
                     errno = ENOLINK;
                     return -1;
                 }
+
+                final_size = (size_t) stat_after.st_size;
             }
 
             // If we got here, we've been successful, so don't try again.
             break;
         }
     }
 
+    /* Truncate if O_CREAT and the file preexisted. */
     if (trunc)
     {
-        if (ftruncate(currentfd, 0) < 0)
+        /* Do not truncate if the size is already zero, some
+         * filesystems don't support ftruncate() with offset>=size. */
+        assert(final_size != (size_t) -1);
+
+        if (final_size != 0)
         {
-            close(currentfd);
-            return -1;
+            int tr_ret = ftruncate(currentfd, 0);
+            if (tr_ret < 0)
+            {
+                Log(LOG_LEVEL_ERR,
+                    "safe_open: unexpected failure (ftruncate: %s)",
+                    GetErrorStr());
+                close(currentfd);
+                return -1;
+            }
         }
     }
 
```

Comment:
Looks good. You should clone the test case `test_safe_open_safe_switched_symlink`, and call it with `O_TRUNC`. This will verify that your fix works even when facing race conditions. Of course you'll have to compare against empty content instead of the content currently there.